psi4/src/psi4/libfock/DirectDFJK.cc

```diff
@@ -0,0 +1,1785 @@
+
+    C_DGETRI(naux_, cmpq_invp, naux_, pert, &work_p[0], naux_);
+
+//    if (work_p) {
+//        free(work_p);
+//    }
+}
+
+// called after sparsity_prep_pQq() 
+// is called in the memory estimate because the maximum permissable
+// block size is needed for an effective memory estimate, and determining
+// that is equivalent to estimating the total amount of memory
+// Determines the number of blocks and
+void DirectDFJK::prepare_p_blocks() {
+    Block_funcs_.resize(0);
+    Shell_starts_.resize(0);
+    Shell_stops_.resize(0);
+    k_disps_.resize(0);
+
+    size_t current_costs = 0;
+    // block max functions
+    size_t bmf = 0;
+
+    size_t block_size = 0;
+    size_t shell_funcs;
+    size_t biggest_shell = 0;
+
+    // one of two things. Either nelectron/2 + 1 or C_left_ao->ncol()
+    size_t charges_z;
+
+    /* Used because psi won't tell us the number of occupied orbitals until 
+     * the actual scf iterations, so we need to use this as a stand in.
+     * I'm not a huge fan either. */
+    if (C_left_ao_.size() == 0) {
+        // atomic charges float
+        double charges_f;
+
+        /*	for (size_t func_it = 0; func_it < schwarz_func_ints_.size(); func_it++) {
+                        sparse_fpf_ += schwarz_func_ints_[func_it];
+                }	*/
+
+        // We need to figure out the biggest shell in the primary basis set
+        // to allocate memory for AO construction
+        for (size_t i = 0; i < primary_->nshell(); i++) {
+            if (primary_->shell(i).nfunction() > biggest_shell) {
+                biggest_shell = primary_->shell(i).nfunction();
+            }
+        }
+
+        std::shared_ptr<Molecule> mol_ptr = primary_->molecule();
+
+        for (int atom_iter = 0; atom_iter < mol_ptr->natom(); atom_iter++) charges_f += mol_ptr->fZ(atom_iter);
+
+        charges_f = charges_f / 2.0;
+
+        charges_z = static_cast<size_t>(charges_f);
+        ++charges_z;
+    } else {
+        charges_z = C_left_ao_[0]->ncol();
+    }
+    //
+    // D_ao_ and D_
+    current_costs += 3 * nbf_ * nbf_;
+    // Coulomb Metrics
+    current_costs += 2 * naux_ * naux_;
+    current_costs += 4 * nbf_ * nbf_;
+    // adding in costs for U and A
+    current_costs += biggest_shell * naux_ * charges_z;
+    current_costs += biggest_shell * naux_ * charges_z;
+    current_costs += omp_nthread_ * 5 * nbf_;
+
+    if (memory_ - current_costs > memory_) {
+        std::stringstream error_message;
+        error_message << "DirectDFJK: not enough memory for overhead! Overhead requires "
+                      << current_costs * 8 / (1024 * 1024) << " [MiB].";
+        throw PSIEXCEPTION(error_message.str().c_str());
+    }
+
+    Shell_starts_.push_back(0);
+    for (size_t shell_iter = 0; shell_iter < p_shells_; shell_iter++) {
+        shell_funcs = primary_->shell(shell_iter).nfunction();
+        // The 5 is where we're getting our fudge factor we're getting charges_z
```

Comment:
What 5?

Reply:
That comment is for an old version of the code. I will fix it.